pkg/apis/storage/types.go

```diff
@@ -308,6 +309,17 @@ type CSIDriverSpec struct {
 	// more modes may be added in the future.
 	// +optional
 	VolumeLifecycleModes []VolumeLifecycleMode
+
+	// If set to true, storageCapacity indicates that the CSI
+	// volume driver wants pod scheduling to consider the storage
+	// capacity that the driver deployment will report by creating
+	// CSIStorageCapacity objects with capacity information.
+	//
+	// This is an alpha field and only available when the CSIStorageCapacity
+	// feature is enabled. The default is false.
+	//
+	// +optional
+	StorageCapacity *bool
```

Comment:
As CSI grows more and more optional capabilities, do we want the API to evolve to a list of specific bool fields?  Or do we want an enumerated "capability list"?  

In general, we try to eschew `bool`, but it seems appropriate for these cases.  If we keep that, maybe we want to sub-struct here?  e.g. `capabilites.storageCapacity` so that would expand more cleanly when we inevitably have a second, third, fourth bool?

Reply:
So I can ask the same question again - if there will one day be several of these, do they want to be in their own struct?

e.g.

```
type CSIDriver {
  //...
  StorageCapacity *StorageCapacityParams
  OtherFeature *OtherFeatureParams
  YetAnother *YetAnotherParams
}
```

vs

```
type CSIDriver {
  //...
  Features *CSIDriverFeatures
}

type CSIDriverFeatures {
  StorageCapacity *StorageCapacityParams
  OtherFeature *OtherFeatureParams
  YetAnother *YetAnotherParams
}
```

Something is still optional...

Reply:
We actually do have 2 more features being added to CSIDriver related to fsgroup and selinux. Although right now, every parameter in CSIDriver is about a feature. I'm not sure what we would put alongside `Features` in the top level. cc @jsafrane

Reply:
For the scenario where we may want to consider different modes of operation, then the guidance is to use a string instead of a bool so that we can expand to new behaviors in the future.

Reply:
So is the conclusion that `StorageCapacity` should be a string with (currently) two values, `StorageCapacityNone = ""` (the default) and `StorageCapacityCheck = Check` for "driver provides storage capacity, check it during provisioning"?

Reply:
Or perhaps we need to reserve the empty string for `StorageCapacityUnknown` for drivers that don't specify anything explicitly? The new values then are `StorageCapacityCheck = Check` and `StorageCapacityIgnore = Ignore`. Suggestions for better names welcome...

Reply:
two parts:

1) Should we have a sub-struct?  I am fine to not have one if you think this is consistent with other fields in the type. Your call.

2) Bool or enum?  I generally try to avoid bools, but I am not close enough to this to make an argument about whether this and other fields are likely to have 3rd options, and whether string-ness has negative impact on the API.

I do prefer explicit values ("NotSupported" or "Ignore") to empty strings and the defaulting logic supports that.

Reply:
I don't think we need a sub-struct, so let's rule that out.

I like strings better and they should work equally well. My only concern is that it is then inconsistent with the rest of the `CSIDriver` API where pointer to bool was used. I'll let @msau42 decide - strings or bool? If I don't hear something, then I'll switch to strings.

> I do prefer explicit values ("NotSupported" or "Ignore") to empty strings and the defaulting logic supports that.

That means directly embedding a string (not a pointer to a string), right?

I prefer "storageCapacity: Ignore" (default) and "storageCapacity: Check", with `StorageCapacityIgnore` and `StorageCapacityCheck` as names for the constants in Go.

"NotSupported" is less accurate because it might be supported and just not complete yet (i.e. in the rollout phase). If anyone cares about that distinction, then "NotSupported" could become a third value 
in the future.

Or perhaps add it already now? It's not a big difference and feels more natural than "Ignore" for a driver that truly doesn't support the feature. So many choices...

Reply:
Do you think we will have a 3rd mode of operation in the future? "Check", I think is too broad. Any future type would also probably involving checking, but perhaps with some different semantic.

Reply:
So what would you prefer? Perhaps "StorageCapacity: Ignore" (default) and "StorageCapacity: RequiredForPodScheduling"?

Reply:
We discussed offline and can't think of a scenario where we would want to add a 3rd mode in the future. The potential future use cases we thought about, max volume capacity, and total capacity, could be accomplished by adding new fields to the capacity object, but doesn't need to be distinguished in the csidriver object.

So I think a bool is fine